docs/proposals/20230407-flexible-managed-k8s-endpoints.md

```diff
@@ -0,0 +1,519 @@
+
+#### Story 2
+
+As a cluster operator, I want to be able to provision both "unmanaged" and "managed" Kubernetes clusters from the same management cluster, so that I can support different requirements and use cases as needed whilst using a single operating model.
+
+#### Story 3
+
+As a Cluster API provider implementor, I want to be able to return the control plane endpoint created by the `ControlPlane Provider`, so that it fits naturally with how most of the native Managed Kubernetes implementations works.
+
+#### Story 4
+
+As a Cluster API provider developer, I want guidance on how to incorporate a managed Kubernetes service into my provider, so that its usage is compatible with Cluster API architecture/features and its usage is consistant with other providers.
+
+#### Story 5
+
+As a Cluster API provider developer, I want to enable the ClusterClass feature for a Managed Kubernetes service, so that users can take advantage of an improved UX with ClusterClass-based clusters.
+
+#### Story 6
+
+As a cluster operator, I want to use Cluster API to provision and manage the lifecycle of worker nodes that utilizes my cloud providers' managed instances (if they support them), so that I don't have to worry about the management of these instances.
+
+#### Story 7
+
+As a service provider I want to be able to offer Managed Kubernetes clusters by using CAPI referencing my own managed control plane implementation that satisfies Cluster API contracts.
+
+### Design
+
+Below we are documenting API changes and controllers changes required to omit the `<Infra>Cluster` entirely and to allow the `ControlPlane Provider` component  to take ownership of the responsibility of creating the control plane endpoint.
+
+#### Core Cluster API changes
+
+This proposal does not introduce any breaking changes for the existing "core" API. More specifically:
+
+The existing Cluster API types are already able to omit the `<Infra>Cluster`:
+
+- The `infrastructureRef` field on the Cluster object is already a pointer and thus it could be set to nil, and in fact we are already creating Clusters without `infrastructureRef` when we use a cluster class).
+- The `infrastructure.Ref` field on the ClusterClass objects already a pointer and thus it could be set to nil, but in this case it is required to change the validation webhook to allow the user to not specify it; on top of that, when validating inline patches, we should reject patches targeting the infrastructure template objects if not specified.
+
+In order to allow the `ControlPlane Provider` component  to take ownership of the responsibility of creating the control plane endpoint we are going to introduce a new `ClusterEndpoint` CRD, below some example:
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterEndpoint
```

Comment:
I suggested ClusterEndpoint, given that we can eventually expand it with a `type` field of sorts to determine what it applies to

Reply:
+1 to keep it generic, so in the future this could cover also a use case different from the current ControlPlaneEndpoint (e.g internal + external control plane endpoints).

Reply:
thanks everyone for weighing, is that sufficient to address your concerns @CecileRobertMichon?

Reply:
Super, thanks!